fix(consumer): missed consumer update due to wrong version in cache

Revolyssup opened the pull request:
### Description
Fixes https://github.com/apache/apisix/issues/12393
This PR introduced a critical bug https://github.com/apache/apisix/pull/11840 due to which consumers are not updated  until a subsequent update in credentials due to credential version being used even for checking consumer update.

More context on this comment: https://github.com/apache/apisix/pull/11840/files#r2193898420

### Checklist

- [x] I have explained the need for this PR and the problem it solves
- [x] I have explained the changes or the new features added to this PR
- [x] I have added tests corresponding to this change
- [ ] I have updated the documentation to reflect this change
- [x] I have verified that this change is backward compatible (If not, please discuss on the [APISIX mailing list](https://github.com/apache/apisix/tree/master#community) first)

<!--

Note

1. Mark the PR as draft until it's ready to be reviewed.
2. Always add/update tests for any changes unless you have a good reason.
3. Always update the documentation to reflect the changes made in the PR.
4. Make a new commit to resolve conversations instead of `push -f`.
5. To resolve merge conflicts, merge master instead of rebasing.
6. Use "request review" to notify the reviewer after making changes.
7. Only a reviewer can mark a conversation as resolved.

-->

nic-6443 commented on apisix/consumer.lua:
```diff
@@ -159,17 +120,49 @@ function plugin_consumer()
+                    local consumer_name = get_consumer_name_from_credential_etcd_key(val.key)
+                    local the_consumer = consumers:get(consumer_name)
+                    if the_consumer and the_consumer.value then
+                        consumer = consumers_id_lrucache(the_consumer.value, val.modifiedIndex+the_consumer.modifiedIndex,
```
we should concatenate those `modifiedIndex` in string way.

nic-6443 replied:
lru key should be `val.key`

nic-6443 commented on apisix/consumer.lua:
```diff
@@ -159,17 +120,49 @@ function plugin_consumer()
+                        goto CONTINUE
+                    end
+                else
+                    consumer = consumers_id_lrucache(val.value, val.modifiedIndex,
```
```suggestion
                    consumer = consumers_id_lrucache(val.key, val.modifiedIndex,
```

nic-6443 commented on apisix/consumer.lua:
```diff
@@ -159,17 +120,49 @@ function plugin_consumer()
+                    local consumer_name = get_consumer_name_from_credential_etcd_key(val.key)
+                    local the_consumer = consumers:get(consumer_name)
+                    if the_consumer and the_consumer.value then
+                        consumer = consumers_id_lrucache(val.value.id .. name, val.modifiedIndex..the_consumer.modifiedIndex,
```
why not use `val.key` ?

Baoyuantop commented:
Will this pr fixed https://github.com/apache/apisix/issues/12393?

Revolyssup commented:
> Will this pr fixed #12393?

Yes

bzp2010 commented:
Truth be told, I didn't understand your modifications. Can you provide more insight? 🤔

Revolyssup commented:
> * change is backward compatible (If not, please discuss on the [APISIX mailing list](https://github.com/apache/apisix/tree/master#community) first)

As this comment explains https://github.com/apache/apisix/pull/11840/files#r2193898420
The `version` being used in the cache was that of credential modified index. This means until the credential is modified, we will use stale value from cache. 
But in the scenario, where the underlying consumer has been modified, we need to recalculate and not use the old value. Therefore in my changes, I have used `credential.modified..consumer.modified` as the version, so that even when underlying consumer changes, a resync is triggered.

bzp2010 commented:
Thanks for the explanation, I seem to have run into some GitHub bug that prevents the web page from pinpointing the exact comment you pointed out.
Anyway, I've seen it. I will check the code again.
